packages/ciphernode/net/src/network_peer.rs

```diff
@@ -29,135 +51,254 @@ pub struct NodeBehaviour {
+                    info!("Failed to dial {peer_id:?}: {error}");
+                    if let Some(dialer) = &mut self.dialer {
+                        dialer.do_send(NetworkPeerEvent::OutgoingConnectionError {
+                            connection_id,
+                            error: Arc::new(error),
+                        });
+                    }
+                }
+
+                SwarmEvent::IncomingConnectionError { error, .. } => {
+                    warn!("{:#}", anyhow::Error::from(error))
+                }
+
+                SwarmEvent::Behaviour(NodeBehaviourEvent::Kademlia(e)) => {
+                    debug!("Kademlia event: {:?}", e);
+                }
+
+                SwarmEvent::Behaviour(NodeBehaviourEvent::Mdns(mdns::Event::Discovered(list))) => {
+                    for (peer_id, _multiaddr) in list {
+                        trace!("mDNS discovered a new peer: {peer_id}");
+                        self.swarm
+                            .behaviour_mut()
+                            .gossipsub
+                            .add_explicit_peer(&peer_id);
+                    }
+                }
+
+                SwarmEvent::Behaviour(NodeBehaviourEvent::Mdns(mdns::Event::Expired(list))) => {
+                    for (peer_id, _multiaddr) in list {
+                        trace!("mDNS discover peer has expired: {peer_id}");
+                        self.swarm
+                            .behaviour_mut()
+                            .gossipsub
+                            .remove_explicit_peer(&peer_id);
+                    }
+                }
+
+                SwarmEvent::Behaviour(NodeBehaviourEvent::Gossipsub(
+                    gossipsub::Event::Message {
+                        propagation_source: peer_id,
+                        message_id: id,
+                        message,
+                    },
+                )) => {
+                    trace!("Got message with id: {id} from peer: {peer_id}",);
+                    mgr.do_send(NetworkPeerEvent::GossipData(message.data));
+                }
+
+                SwarmEvent::NewListenAddr { address, .. } => {
+                    info!("Local node is listening on {address}");
+                }
+                _ => {}
+            };
+        }
+        Ok(())
     }
+}
+
+impl Actor for NetworkPeer {
+    type Context = Context<Self>;
+
+    fn started(&mut self, ctx: &mut Self::Context) {
+        // Setup stream handling for swarm events
+        ctx.add_stream(Box::pin(futures::stream::unfold(
+            &mut self.swarm,
+            |swarm| async move { Some((swarm.select_next_some().await, swarm)) },
+        )));
 
-    pub fn rx(&mut self) -> broadcast::Receiver<NetworkPeerEvent> {
-        self.event_tx.subscribe()
+        if let Some(dialer) = &mut self.dialer {
+            dialer.send(SetNetworkPeer(ctx.address()));
+            if !self.peers.is_empty() {
+                dialer.do_send(DialPeers(self.peers.clone()));
```

Comment:
It might be simpler to spawned `self.peers.len()` dialers for each peer in the peer list with a dialer being a short lived actor in charge of dialing a specific address and handling expbackoff etc. having the dialer handle multiple requests seems more complex than it needs to be?

Reply:
The issue with that approach is passing events between the dialer and the peer. In that case, we would have to manage all the dialers inside the NetworkPeer and loop through the dialers for swarm events to send the event to the appropriate dialer. Although, we could create a separate EventBus for NetworkPeerEvent that the dialers can subscribe to in order to receive the events.

```mermaid
graph TD
    subgraph NetworkPeerActor
        NPA[NetworkPeerActor]
    end

    subgraph DialerActors
        DA1[DialerActor 1]
        DA2[DialerActor 2]
        DA3[DialerActor 3]
    end

    subgraph EventBus
        EB[EventBus]
    end

    %% Connections
    NPA --> EB
    EB --> DA1
    EB --> DA2
    EB --> DA3
```